File: android/src/main/kotlin/com/zcreations/executorch_flutter/ExecutorchModelManager.kt
/**
 * ExecuTorch Model Manager - Core Android ExecuTorch Integration
 *
 * This class provides the core ExecuTorch integration for the Android platform,
 * implementing model lifecycle management, inference execution, and tensor data conversion.
 * It integrates with the latest ExecuTorch Android AAR 0.7.0+ using the Module.load() API pattern.
 *
 * Features:
 * - ExecuTorch Module.load() integration with proper exception handling
 * - Thread-safe model management with concurrent access support
 * - Efficient tensor data conversion between Flutter and ExecuTorch formats
 * - Memory management with proper resource cleanup
 * - Performance monitoring and comprehensive error handling
 * - Model metadata extraction and validation
 *
 * Integration Pattern:
 * - Uses org.pytorch.executorch.Module for model loading and inference
 * - EValue/Tensor API for input/output data handling
 * - Proper Android context awareness and lifecycle management
 * - Background thread execution for non-blocking operations
 */
package com.zcreations.executorch_flutter

import android.content.Context
import android.util.Log
import kotlinx.coroutines.*
import org.pytorch.executorch.EValue
import org.pytorch.executorch.Module
import org.pytorch.executorch.Tensor
import java.io.File
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import kotlin.collections.HashMap

// Import Pigeon generated API
import com.zcreations.executorch_flutter.generated.*

/**
 * Manages ExecuTorch model lifecycle and inference operations
 */
class ExecutorchModelManager(
    private val context: Context
) {
    companion object {
        private const val TAG = "ExecutorchModelManager"
        private const val MODEL_ID_PREFIX = "executorch_model_"
    }

    // Model storage
    private val loadedModels = ConcurrentHashMap<String, LoadedModel>()
    private val modelCounter = java.util.concurrent.atomic.AtomicInteger(0)
    private var isDebugLoggingEnabled = false

    /**
     * Represents a loaded ExecuTorch model
     */
    private data class LoadedModel(
        val module: Module,
        val filePath: String,
        val loadTime: Long = System.currentTimeMillis()
    )

    /**
     * Load an ExecuTorch model from a file path
     * Throws ModelLoadException on failure
     */
    suspend fun load(filePath: String): ModelLoadResult = withContext(Dispatchers.IO) {
        if (isDebugLoggingEnabled) {
            Log.d(TAG, "Loading ExecuTorch model from: $filePath")
        }

        // Validate file exists and is readable
        val file = File(filePath)
        if (!file.exists()) {
            throw ModelLoadException("Model file not found: $filePath")
        }

        if (!file.canRead()) {
            throw ModelLoadException("Model file not readable: $filePath")
        }

        // Generate unique model ID
        val modelId = generateModelId()

        try {
            // Load model using ExecuTorch Module.load() API
            if (isDebugLoggingEnabled) {
                Log.d(TAG, "Loading ExecuTorch module with Module.load()")
            }
            val module = Module.load(filePath)

            // Store loaded model
            val loadedModel = LoadedModel(
                module = module,
                filePath = filePath
            )

            loadedModels[modelId] = loadedModel

            if (isDebugLoggingEnabled) {
                Log.d(TAG, "Successfully loaded model: $modelId from $filePath")
            }

            ModelLoadResult(modelId = modelId)

        } catch (e: Exception) {
            // Clean up on failure
            loadedModels.remove(modelId)

            Log.e(TAG, "Failed to load ExecuTorch module: $filePath", e)
            throw ModelLoadException("ExecuTorch module loading failed: ${e.message}", e)
        }
    }

    /**
     * Run forward pass (inference) on a loaded model
     * Returns output tensors directly
     * Throws InferenceException on failure
     */
    suspend fun forward(modelId: String, inputs: List<TensorData?>): List<TensorData?> = withContext(Dispatchers.Default) {
        val loadedModel = loadedModels[modelId]
            ?: throw ModelNotFoundException("Model not found: $modelId")

        if (isDebugLoggingEnabled) {
            Log.d(TAG, "Running inference on model: $modelId")
        }

        try {
            // Filter out null inputs
            val validInputs = inputs.filterNotNull()

            // Convert Flutter tensors to ExecuTorch EValues
            val inputEValues = convertTensorsToEValues(validInputs)

            val startTime = System.nanoTime()

            // Execute inference using ExecuTorch Module.forward()
            val outputs = loadedModel.module.forward(*inputEValues.toTypedArray())

            val endTime = System.nanoTime()
            val executionTimeMs = (endTime - startTime) / 1_000_000.0

            // Convert outputs back to Flutter tensors
            val outputTensors = convertEValuesToTensors(outputs)

            if (isDebugLoggingEnabled) {
                Log.d(TAG, "Inference completed in ${executionTimeMs}ms for model: $modelId")
            }

            outputTensors

        } catch (e: Exception) {
            Log.e(TAG, "Inference failed for model: $modelId", e)
            throw InferenceException("Inference failed: ${e.message}", e)
        }
    }

    /**
     * Dispose a loaded model and free its resources
     * Throws ModelNotFoundException if model not found
     */
    suspend fun dispose(modelId: String) = withContext(Dispatchers.IO) {
        val loadedModel = loadedModels.remove(modelId)
            ?: throw ModelNotFoundException("Model not found: $modelId")

        // Note: ExecuTorch Module cleanup handled by garbage collection
        if (isDebugLoggingEnabled) {
            Log.d(TAG, "Disposed model: $modelId")
        }
    }

    /**
     * Get list of loaded model IDs
     */
    fun getLoadedModels(): List<String?> {
        return loadedModels.keys.toList()
    }

    /**
     * Dispose all loaded models
     */
    suspend fun disposeAllModels() = withContext(Dispatchers.IO) {
        val modelIds = loadedModels.keys.toList()
        modelIds.forEach { modelId ->
            dispose(modelId)
        }
        if (isDebugLoggingEnabled) {
            Log.d(TAG, "Disposed all models (${modelIds.size} total)")
        }
    }

    /**
     * Enable or disable debug logging
     */
    fun setDebugLogging(enabled: Boolean) {
        isDebugLoggingEnabled = enabled
        if (isDebugLoggingEnabled) {
            Log.d(TAG, "Debug logging enabled")
        }
    }

    // Private helper methods

    private fun generateModelId(): String {
        return MODEL_ID_PREFIX + UUID.randomUUID().toString().replace("-", "").substring(0, 8)
    }

    private fun convertTensorsToEValues(tensors: List<TensorData>): List<EValue> {
        return tensors.map { tensorData ->
            val tensor = when (tensorData.dataType) {
                TensorType.FLOAT32 -> {
                    val buffer = ByteBuffer.wrap(tensorData.data)
                        .order(ByteOrder.nativeOrder())
                        .asFloatBuffer()
                    val floatArray = FloatArray(buffer.remaining())
                    buffer.get(floatArray)

                    Tensor.fromBlob(
                        floatArray,
                        tensorData.shape.filterNotNull().map { it.toLong() }.toLongArray()
                    )
                }

                TensorType.INT32 -> {
                    val buffer = ByteBuffer.wrap(tensorData.data)
                        .order(ByteOrder.nativeOrder())
                        .asIntBuffer()
                    val intArray = IntArray(buffer.remaining())
                    buffer.get(intArray)

                    Tensor.fromBlob(
                        intArray,
                        tensorData.shape.filterNotNull().map { it.toLong() }.toLongArray()
                    )
                }

                TensorType.INT8, TensorType.UINT8 -> {
                    Tensor.fromBlob(
                        tensorData.data,
                        tensorData.shape.filterNotNull().map { it.toLong() }.toLongArray()
                    )
                }
            }

            EValue.from(tensor)
        }
    }

    private fun convertEValuesToTensors(eValues: Array<EValue>): List<TensorData?> {
        return eValues.mapIndexed { index, eValue ->
            val tensor = eValue.toTensor()
            val shape = tensor.shape().map { it.toInt() }

            // Determine data type and convert accordingly
            val (dataType, data) = when {
                tensor.dataAsFloatArray != null -> {
                    val floatArray = tensor.dataAsFloatArray
                    val buffer = ByteBuffer.allocate(floatArray.size * 4)
                        .order(ByteOrder.nativeOrder())
                    buffer.asFloatBuffer().put(floatArray)
                    TensorType.FLOAT32 to buffer.array()
                }

                tensor.dataAsIntArray != null -> {
                    val intArray = tensor.dataAsIntArray
                    val buffer = ByteBuffer.allocate(intArray.size * 4)
                        .order(ByteOrder.nativeOrder())
                    buffer.asIntBuffer().put(intArray)
                    TensorType.INT32 to buffer.array()
                }

                else -> {
                    // Default to byte array
                    TensorType.UINT8 to tensor.dataAsByteArray
                }
            }

            TensorData(
                shape = shape.map { it.toLong() },
                dataType = dataType,
                data = data,
                name = "output_$index"
            )
        }
    }

}

File: darwin/Sources/executorch_flutter/ExecutorchModelManager.swift
/**
 * ExecuTorch Model Manager - Core iOS ExecuTorch Integration
 *
 * This class provides the core ExecuTorch integration for the iOS platform,
 * implementing model lifecycle management, inference execution, and tensor data conversion.
 * It integrates with ExecuTorch iOS frameworks using the Module/Tensor/Value API pattern.
 *
 * Features:
 * - ExecuTorch Module integration with proper error handling
 * - Thread-safe model management
 * - Efficient tensor data conversion between Flutter and ExecuTorch formats
 * - User-controlled memory management and resource cleanup
 * - Performance monitoring and comprehensive error handling
 * - Model metadata extraction and validation
 * - Memory mapping support for large models
 *
 * Integration Pattern:
 * - Uses ExecuTorchModule for model loading and inference
 * - ExecuTorchTensor/ExecuTorchValue API for input/output data handling
 * - Async/await for non-blocking operations
 * - User-controlled model lifecycle
 *
 * Requirements:
 * - iOS 13.0+ for ExecuTorch framework support
 * - ExecuTorch.framework linked and embedded
 * - Swift 5.9+ for async/await support
 */
import Foundation

#if os(iOS)
import Flutter
#elseif os(macOS)
import FlutterMacOS
#endif

// ExecuTorch only supports arm64 architecture on macOS
#if os(macOS)
  #if arch(arm64)
    import ExecuTorch
  #else
    #error("ExecuTorch only supports arm64 (Apple Silicon) on macOS. Intel Macs (x86_64) are NOT supported. Add to macos/Runner/Configs/Release.xcconfig: ARCHS = arm64 and ONLY_ACTIVE_ARCH = YES")
  #endif
#else
  import ExecuTorch
#endif

/**
 * Actor-based model manager for thread-safe ExecuTorch operations
 */
actor ExecutorchModelManager {

    internal static let TAG = "ExecutorchModelManager"
    private static let MODEL_ID_PREFIX = "executorch_model_"

    // Model storage
    private var loadedModels: [String: LoadedModel] = [:]
    private var modelCounter: Int = 0

    // Debug logging control
    private var isDebugLoggingEnabled = false

    /**
     * Represents a loaded ExecuTorch model with metadata
     */
    private struct LoadedModel {
        let module: Module
        let filePath: String
        let loadTime: TimeInterval = Date().timeIntervalSince1970
    }

    init() {
        // Note: log() cannot be used here because init is synchronous
        // and we want to log initialization regardless of debug setting
        print("[\(Self.TAG)] ExecutorchModelManager initialized")
    }

    deinit {
        // Note: deinit is synchronous, so we use print directly
        if isDebugLoggingEnabled {
            print("[\(Self.TAG)] ExecutorchModelManager deinitialized")
        }
    }

    // MARK: - Public API

    /**
     * Load an ExecuTorch model from a file path
     * Throws ExecutorchError on failure
     */
    func load(filePath: String) async throws -> ModelLoadResult {
        log("Loading ExecuTorch model from: \(filePath)")

        // Validate file exists and is readable
        guard FileManager.default.fileExists(atPath: filePath) else {
            logError("File not found: \(filePath)")
            throw ExecutorchError.fileNotFound(filePath)
        }

        guard FileManager.default.isReadableFile(atPath: filePath) else {
            logError("File not readable: \(filePath)")
            throw ExecutorchError.modelLoadFailed(filePath, nil)
        }

        // Generate unique model ID
        let modelId = generateModelId()

        do {
            // Load model using ExecuTorch Module API
            log("Creating Module with file path")
            let module = Module(filePath: filePath)

            // Load the forward method (most common case)
            log("Loading 'forward' method")
            try module.load("forward")

            // Verify the module is loaded
            guard module.isLoaded("forward") else {
                logError("Module.isLoaded('forward') returned false for: \(filePath)")
                throw ExecutorchError.modelLoadFailed(filePath, nil)
            }

            // Store loaded model
            let loadedModel = LoadedModel(
                module: module,
                filePath: filePath
            )

            loadedModels[modelId] = loadedModel

            log("Successfully loaded model: \(modelId) from \(filePath)")

            return ModelLoadResult(modelId: modelId)

        } catch {
            // Clean up on failure
            loadedModels.removeValue(forKey: modelId)

            logError("Failed to load ExecuTorch module: \(filePath), error: \(error)")
            throw ExecutorchError.modelLoadFailed(filePath, error)
        }
    }

    /**
     * Run forward pass (inference) on a loaded model
     * Throws ExecutorchError on failure
     * Returns output tensors directly
     */
    func forward(modelId: String, inputs: [TensorData?]) async throws -> [TensorData?] {
        guard let loadedModel = loadedModels[modelId] else {
            logError("Model not found: \(modelId)")
            throw ExecutorchError.modelNotFound(modelId)
        }

        log("Running forward pass on model: \(modelId)")

        // Convert Flutter tensors to ExecuTorch Values
        let validInputs = inputs.compactMap { $0 }
        let inputValues = try convertTensorsToValues(validInputs)

        let startTime = CFAbsoluteTimeGetCurrent()

        do {
            // Execute inference using ExecuTorch Module.forward()
            let outputs = try loadedModel.module.forward(inputValues)

            let endTime = CFAbsoluteTimeGetCurrent()
            let executionTimeMs = (endTime - startTime) * 1000

            // Convert outputs back to Flutter tensors
            let outputTensors = try convertValuesToTensors(outputs)

            log("Forward pass completed in \(executionTimeMs)ms for model: \(modelId)")

            return outputTensors

        } catch {
            logError("Forward pass failed for model: \(modelId), error: \(error)")
            throw ExecutorchError.inferenceFailed(modelId, error)
        }
    }

    /**
     * Dispose a loaded model and free its resources
     * Throws ExecutorchError.modelNotFound if model not found
     */
    func dispose(modelId: String) async throws {
        let loadedModel = loadedModels.removeValue(forKey: modelId)

        guard loadedModel != nil else {
            logError("Cannot dispose - model not found: \(modelId)")
            throw ExecutorchError.modelNotFound(modelId)
        }

        // Note: ExecuTorchModule cleanup handled by ARC
        log("Disposed model: \(modelId)")
    }

    /**
     * Get list of loaded model IDs
     */
    func getLoadedModels() throws -> [String?] {
        return Array(loadedModels.keys)
    }

    /**
     * Enable or disable debug logging
     */
    func setDebugLogging(enabled: Bool) throws {
        isDebugLoggingEnabled = enabled
        log("Debug logging \(enabled ? "enabled" : "disabled")")
    }

    // MARK: - Private Helper Methods

    /**
     * Log a debug message (only if debug logging is enabled)
     */
    private func log(_ message: String) {
        if isDebugLoggingEnabled {
            print("[\(Self.TAG)] \(message)")
        }
    }

    /**
     * Log an error message (always logged regardless of debug setting)
     */
    private func logError(_ message: String) {
        NSLog("[\(Self.TAG)] ERROR: \(message)")
    }

    private func generateModelId() -> String {
        modelCounter += 1
        return "\(Self.MODEL_ID_PREFIX)\(UUID().uuidString.prefix(8))"
    }

    private func convertTensorsToValues(_ tensors: [TensorData]) throws -> [Value] {
        return try tensors.map { tensorData in
            let shape = tensorData.shape.compactMap { $0 }.map { Int($0) }

            switch tensorData.dataType {
            case .float32:
                // Convert bytes back to float array
                let bytes = tensorData.data.data
                let floatCount = bytes.count / 4
                var floats = [Float](repeating: 0, count: floatCount)
                _ = floats.withUnsafeMutableBytes { floatBytes in
                    bytes.copyBytes(to: floatBytes)
                }

                let tensor = Tensor(floats, shape: shape)
                return Value(tensor)

            case .int32:
                // Convert bytes back to int32 array
                let bytes = tensorData.data.data
                let intCount = bytes.count / 4
                var ints = [Int32](repeating: 0, count: intCount)
                _ = ints.withUnsafeMutableBytes { intBytes in
                    bytes.copyBytes(to: intBytes)
                }

                let tensor = Tensor(ints, shape: shape)
                return Value(tensor)

            case .int8, .uint8:
                // Use bytes directly
                let bytes = tensorData.data.data
                let byteArray = Array(bytes)
                let tensor = Tensor(byteArray, shape: shape)
                return Value(tensor)
            }
        }
    }

    private func convertValuesToTensors(_ values: [Value]) throws -> [TensorData] {
        return try values.enumerated().map { (index, value) in
            // Try to get tensor as Float first (most common), then Int32, then Byte
            if let tensor: Tensor<Float> = value.tensor() {
                let shape = tensor.shape.map { Int64($0) }
                let floatArray = tensor.withUnsafeBytes { (buffer: UnsafeBufferPointer<Float>) -> [Float] in
                    Array(buffer)
                }
                let data = Data(bytes: floatArray, count: floatArray.count * MemoryLayout<Float>.size)
                return TensorData(
                    shape: shape,
                    dataType: .float32,
                    data: FlutterStandardTypedData(bytes: data),
                    name: "output_\(index)"
                )
            } else if let tensor: Tensor<Int32> = value.tensor() {
                let shape = tensor.shape.map { Int64($0) }
                let intArray = tensor.withUnsafeBytes { (buffer: UnsafeBufferPointer<Int32>) -> [Int32] in
                    Array(buffer)
                }
                let data = Data(bytes: intArray, count: intArray.count * MemoryLayout<Int32>.size)
                return TensorData(
                    shape: shape,
                    dataType: .int32,
                    data: FlutterStandardTypedData(bytes: data),
                    name: "output_\(index)"
                )
            } else if let tensor: Tensor<UInt8> = value.tensor() {
                let shape = tensor.shape.map { Int64($0) }
                let byteArray = tensor.withUnsafeBytes { (buffer: UnsafeBufferPointer<UInt8>) -> [UInt8] in
                    Array(buffer)
                }
                let data = Data(byteArray)
                return TensorData(
                    shape: shape,
                    dataType: .uint8,
                    data: FlutterStandardTypedData(bytes: data),
                    name: "output_\(index)"
                )
            } else {
                throw ExecutorchError.inferenceFailed("Output \(index) has unsupported tensor type", nil)
            }
        }
    }

}